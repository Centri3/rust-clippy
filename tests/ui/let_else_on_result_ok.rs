//@aux-build:proc_macros.rs
#![allow(clippy::redundant_pattern_matching, unused)]
#![warn(clippy::let_else_on_result_ok)]

#[macro_use]
extern crate proc_macros;

struct A(Result<&'static A, ()>);

enum AnEnum {
    A(Result<&'static A, ()>),
}

struct B;

struct C(u32);

enum D {
    A,
    B,
}

enum E {
    A,
    B(u32),
}

#[repr(C)]
union F {
    a: u32,
}

#[non_exhaustive]
struct G {}

#[non_exhaustive]
enum H {}

fn a() -> Result<(), ()> {
    Ok(())
}

fn b() -> Result<(), ()> {
    let (Ok(_), 1) = (a(), 1) else {
        return Err::<(), _>(());
    };
    Ok(())
}

fn c() -> Result<(), C> {
    todo!()
}

fn d() -> Result<(), D> {
    todo!()
}

fn e() -> Result<(), E> {
    todo!()
}

fn f() -> Result<(), F> {
    todo!()
}

fn g() -> Result<(), G> {
    todo!()
}

fn h() -> Result<(), H> {
    todo!()
}

fn a_constructor() -> A {
    todo!();
}

fn an_enum_constructor() -> AnEnum {
    todo!();
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Lint
    let Ok(_) = c() else {
        return Ok(());
    };
    let Ok(_) = e() else {
        return Ok(());
    };
    let Ok(_) = f() else {
        return Ok(());
    };
    let Ok(_) = g() else {
        return Ok(());
    };
    let Ok(_) = h() else {
        return Ok(());
    };
    // Don't lint
    loop {
        let Ok(_) = c() else {
            continue;
        };
    }
    let Err(_) = a() else {
        return Ok(());
    };
    let Ok(_) = a() else {
        return Ok(());
    };
    let Ok(_) = b() else {
        return Ok(());
    };
    let Ok(_) = d() else {
        return Ok(());
    };
    match a() {
        Ok(a) => a,
        Err(e) => eprintln!("{e:#?}"),
    };
    external! {
        let Ok(_) = a() else {
            return Ok(());
        };
    }
    Ok(())
}

fn no_result_main() {
    // Don't lint
    let Ok(_) = c() else {
        return;
    };
    let Ok(_) = e() else {
        return;
    };
    let Ok(_) = f() else {
        return;
    };
}
