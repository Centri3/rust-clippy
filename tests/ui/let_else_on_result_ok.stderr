error: usage of `let...else` on `Ok`
  --> $DIR/let_else_on_result_ok.rs:84:9
   |
LL |     let Ok(_) = c() else {
   |         ^^^^^
   |
   = note: this will ignore the contents of the `Err` variant
   = help: consider using a `match` instead, or propagating it to the caller
   = note: `-D clippy::let-else-on-result-ok` implied by `-D warnings`

error: usage of `let...else` on `Ok`
  --> $DIR/let_else_on_result_ok.rs:87:9
   |
LL |     let Ok(_) = e() else {
   |         ^^^^^
   |
   = note: this will ignore the contents of the `Err` variant
   = help: consider using a `match` instead, or propagating it to the caller

error: usage of `let...else` on `Ok`
  --> $DIR/let_else_on_result_ok.rs:90:9
   |
LL |     let Ok(_) = f() else {
   |         ^^^^^
   |
   = note: this will ignore the contents of the `Err` variant
   = help: consider using a `match` instead, or propagating it to the caller

error: usage of `let...else` on `Ok`
  --> $DIR/let_else_on_result_ok.rs:93:9
   |
LL |     let Ok(_) = g() else {
   |         ^^^^^
   |
   = note: this will ignore the contents of the `Err` variant
   = help: consider using a `match` instead, or propagating it to the caller

error: usage of `let...else` on `Ok`
  --> $DIR/let_else_on_result_ok.rs:96:9
   |
LL |     let Ok(_) = h() else {
   |         ^^^^^
   |
   = note: this will ignore the contents of the `Err` variant
   = help: consider using a `match` instead, or propagating it to the caller

error: aborting due to 5 previous errors

